### DISCUSSION TITLE
Support `raise_for_status` on request methods and clients

### DISCUSSION THREAD

**Discussion by timofurrer** [encode/httpx#2336] [Ideas]
I often have the following verbose pattern in my code:

```python
response = client.get(...)
response.raise_for_status()
data = response.json()
```

... especially in testing code.

For convenience sake I'd love to see something like this:

```python
response = client.get(..., raise_for_status=True)
data = response.json()
```

and even:

```python
client = httpx.Client(..., raise_for_status=True)
response = client.get(...)
data = response.json()
```

Because it's rather an easy implementation and I wanted to play around a little bit I've already implemented this here: 

https://github.com/encode/httpx/compare/master...timofurrer:httpx:feature/raise-for-status

I'm already aware of the `event_hooks`, but I suppose the described use case is common enough to dedicate this non-intrusive, intuitive flag to it.

WDYT?

**Comment by ml31415** [comment]
If `raise_for_status` would return the response itself instead of `None`, you could simply write:
```
response = client.get(...).raise_for_status()
data = response.json()
```
Which would probably be nicer than another keyword argument and not change the call interface.

**Reply by timofurrer** [reply]
Sounds like a good solution for me, too ...

**Reply by timofurrer** [reply]
I would still have to do it for every call then.

With the proposed solution above you can also configure the `Client()` with `raise_for_status=True` - which would then apply for all of that clients request calls.

**Reply by sushi2all** [reply]
~~this would be very nice, especially considering the documentation says it can be done...🙄~~

https://www.python-httpx.org/quickstart/

**Reply by lovelydinosaur** [reply]
> this would be very nice, especially considering the documentation says it can be done...🙄

Can you clarify?

**Reply by sushi2all** [reply]
Forgive me, the docs are for v0.27.2, and I was using v0.23.3. I see it's been implemented in 0.25.

My bad 🤦🏼‍♂️🙏🏻

**Comment by ronnyhaase** [comment]
Hi!

Any plans to implement this? Because I also dislike the redundant code. I had to make this change for all my requests.

```python
data = httpx.get(...).json()
```
to
```python
response = httpx.get(...)
response.raise_for_status()
data = response.json()
```
This becomes even more redundant when having multiple requests in one file!

Allow me to mention that I'd prefer a new param for this, that's what I'm used to from other libs. Since it could have a default, it would be backwards compatible anyways.
It's similarly well readable, but easier to understand as a param in my opinion:
```python
# A
httpx.get(..., { ..., raise_for_status: True}).json()

# B
```python
https.get(...).raise_for_status().json()
```

I also would like to discuss if HTTPX should actually raise an exception for status codes >= 400 by default, any thoughts?
In that case an additional param would be mandatory though, but since you're aiming for v1.0 mabye now is the time 😄 .

**Comment by james-mchugh** [comment]
I would like to see this functionality as well. I would like to consolidate my exception handling and formatting of a useful helpful message to a single module for a CLI application I'm building. I'd like this exception handling module to be able to handle both `httpx.ConnectionErrors` and `httpx.HTTPStatusErrors` (my API uses a standardized format for error responses). For it to handle status errors, I would have to call `raise_for_status` everywhere I use httpx, which is prone to programmer errors. I'd be willing to put in a PR for this change if it desired.

**Reply by james-mchugh** [reply]
For anyone looking for this functionality for httpx Clients, it's quite simple to create an `httpx.Client` subclass with this exact functionality. See below

```python
class Client(httpx.Client):
    """Automatically raise HTTPStatusErrors for non-200 responses"""

    def request(
        self,
        *args,
        **kwargs,
    ) -> Response:
        """Send the request and raise an exception for error status responses"""
        response = super().request(*args, **kwargs)
        response.raise_for_status()
        return response

```

If desired, you could easily take this a step further and add a `raise_for_status` option to `__init__` to enable/disable this functionality.

**Comment by chu23465** [comment]
Or you can just do
```python
session = httpx.Client(
    transport=transport, 
    proxy=self.proxy, 
    event_hooks={
    "response": [lambda r, *_, **__: r.raise_for_status(),]
})
```
